### PR TITLE
chore: fix cosine residual calculation

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance-linalg/src/kernels.rs [hunk 1]
```diff
@@ -130,6 +134,45 @@ pub fn normalize<T: Float + Sum>(v: &[T]) -> impl Iterator<Item = T> + '_ {
     v.iter().map(move |&x| x / l2_norm)
 }
 
+fn do_normalize_arrow<T: ArrowPrimitiveType>(arr: &dyn Array) -> Result<ArrayRef>
+where
+    <T as ArrowPrimitiveType>::Native: Float + Sum,
+{
+    let v = arr.as_primitive::<T>();
+    Ok(Arc::new(PrimitiveArray::<T>::from_iter_values(normalize(v.values()))) as ArrayRef)
```
Comment (wjones127) [inline rust/lance-linalg/src/kernels.rs:142]: FYI I plan to fix this soon, but `normalize` doesn't dispatch to the optimized SIMD kernels we have yet.

File: rust/lance-index/src/vector/ivf.rs [hunk 2]
```diff
@@ -306,12 +306,16 @@ impl<T: ArrowFloatType + Dot + L2 + Cosine + 'static> IvfImpl<T> {
...
         let dimension = data.ndim();
         let centroids = self.centroids.data();
-        let data = data.data();
-        let metric_type = self.metric_type;
-
         let num_centroids = centroids.len() / dimension;
-        let num_rows = data.len() / dimension;
+        let num_rows = data.num_rows();
+
+        let (data, metric_type) = if self.metric_type == MetricType::Cosine {
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/ivf.rs:312]: should we move this higher? maybe to `search`?

File: rust/lance-index/src/vector/pq.rs [hunk 3]
```diff
@@ -340,23 +340,7 @@ impl<T: ArrowFloatType + Cosine + Dot + L2 + 'static> ProductQuantizer for Produ
...
-                    ),
-                    location: location!(),
-                })?
-                .as_slice()
-                .chunks(self.dimension)
-                .flat_map(normalize)
-                .collect::<Vec<_>>();
-            let data = T::ArrayType::from(values);
-            FixedSizeListArray::try_new_from_values(data, self.dimension as i32)?
+            normalize_fsl(&fsl)?
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/pq.rs:343]: Should we remove cosine type PQ entirely? And let `IVF` handle normalization and resid?

File: rust/lance-linalg/src/kernels.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance-linalg/src/kernels.rs:142]: 🔥

File: rust/lance-index/src/vector/ivf.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/ivf.rs:312]: ok i can try

Comment (eddyxu) [reply rust/lance-index/src/vector/ivf.rs:312]: THe shuffler is still using this piece separately, tho

File: rust/lance-index/src/vector/ivf.rs [hunk 4]
```diff
@@ -253,17 +253,16 @@ impl<T: ArrowFloatType + Dot + L2 + ArrowPrimitiveType> IvfImpl<T> {
...
+            transforms.push(Arc::new(super::transform::NormalizeTransformer::new(
+                vector_column,
+            )));
+        };
 
         // TODO: add range filter
         let ivf_transform = Arc::new(IvfTransformer::new(
             centroids.clone(),
-            metric_type,
+            MetricType::L2,
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/ivf.rs:265]: what about dot product?

File: rust/lance-index/src/vector/pq.rs (hunk 3 above)
Comment (chebbyChefNEQ) [reply rust/lance-index/src/vector/pq.rs:343]: ^^ can we add a `panic!` that make sure we don't create a cosine PQ?

File: rust/lance-linalg/src/kmeans.rs [hunk 5]
```diff
@@ -402,7 +394,7 @@ where
         let mut best_kmeans = Self::empty(k, dimension, params.metric_type);
         let mut best_stddev = f32::MAX;
 
-        let rng = rand::rngs::SmallRng::from_entropy();
+        let rng = SmallRng::from_entropy();
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/kmeans.rs:397]: nit: add a todo to use seeds

File: rust/lance-linalg/src/kmeans.rs [hunk 6]
```diff
@@ -499,12 +489,7 @@ where
                                 .collect();
                         }
                         MetricType::Cosine => {
-                            let normalized = values
-                                .chunks(dimension)
-                                .flat_map(normalize)
-                                .collect::<Vec<_>>();
-                            return compute_partitions_l2(centroids_array, &normalized, dimension)
-                                .collect();
+                            panic!("KMeans: should not use cosine distance to train kmeans, use L2 instead.");
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/kmeans.rs:492]: add this check in the constructor?

File: rust/lance/src/index/vector/ivf.rs [hunk 7]
```diff
@@ -407,13 +410,20 @@ impl Index for IVFIndex {
 impl VectorIndex for IVFIndex {
     #[instrument(level = "debug", skip_all, name = "IVFIndex::search")]
     async fn search(&self, query: &Query, pre_filter: Arc<PreFilter>) -> Result<RecordBatch> {
-        let partition_ids =
-            self.ivf
-                .find_partitions(&query.key, query.nprobes, self.metric_type)?;
+        let mut query = query.clone();
```
Comment (chebbyChefNEQ) [inline rust/lance/src/index/vector/ivf.rs:413]: how cheap is this copy, can we avoid it for the L2 path?

Comment (eddyxu) [reply rust/lance/src/index/vector/ivf.rs:413]: This is very cheap, the query vector is zero copy (copied by pointer), other than that, there are just 6 other string/int fields.

File: rust/lance-linalg/src/kmeans.rs (hunk 6 above)
Comment (eddyxu) [reply rust/lance-linalg/src/kmeans.rs:492]: The match arm needs a branch to handle `MetricType::Cosine` tho.

File: rust/lance-index/src/vector/ivf.rs (hunk 4 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/ivf.rs]: fixed

File: rust/lance-linalg/src/kmeans.rs (hunk 5 above)
Comment (eddyxu) [reply rust/lance-linalg/src/kmeans.rs:398]: done.

File: rust/lance-index/src/vector/ivf.rs [hunk 8]
```diff
@@ -253,20 +253,17 @@ impl<T: ArrowFloatType + Dot + L2 + ArrowPrimitiveType> IvfImpl<T> {
     ) -> Self {
         let mut transforms: Vec<Arc<dyn Transformer>> = vec![];
 
-        // Re-enable it after search path fixed.
-        // if metric_type == MetricType::Cosine {
-        //     transforms.push(Arc::new(super::transform::NormalizeTransformer::new(
-        //         vector_column,
-        //     )));
-        // };
+        let mt = if metric_type == MetricType::Cosine {
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/ivf.rs:256]: nit: make a macro for this?

File: rust/lance/src/index/vector/pq.rs [hunk 9]
```diff
@@ -263,5 +268,246 @@ impl VectorIndex for PQIndex {
...
+        "starting to compute partitions for PQ training, sample size: {}",
+        training_data.value_length()
+    );
+
+    if metric_type == MetricType::Cosine {
+        info!("Normalize training data for PQ training: Cosine");
+        training_data = normalize_fsl(&training_data)?;
+    }
+
+    println!("PQ Training, ivf: {:?} ", ivf);
```
Comment (chebbyChefNEQ) [inline rust/lance/src/index/vector/pq.rs:354]: log

File: rust/lance/src/index/vector/pq.rs [hunk 10]
```diff
@@ -263,5 +268,246 @@ impl VectorIndex for PQIndex {
...
+        let ivf2 = lance_index::vector::ivf::new_ivf(
+            ivf.centroids.values(),
+            ivf.dimension(),
+            MetricType::L2,
+            vec![],
+            None,
+        )?;
+        span!(Level::INFO, "compute residual for PQ training")
+            .in_scope(|| ivf2.compute_residual(&training_data, None))
+            .await?
```
Comment (chebbyChefNEQ) [inline rust/lance/src/index/vector/pq.rs:369]: I think we still need to make sure we don't residulize for dot.

So, make something like `pq_params.should_residulize`?

**Review by chebbyChefNEQ — APPROVED** [review]
just a few `println!` and pq residual needs to be fixed

File: rust/lance/src/index/vector/pq.rs (hunk 10 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/pq.rs:369]: So the design lets the caller decide whether to run residual or not. If provided IVF, it runs residual PQ.

So PQ is orthogonal to distance type as much as possible.

Comment (chebbyChefNEQ) [reply rust/lance/src/index/vector/pq.rs:366]: but we always residualize currently.

Comment (chebbyChefNEQ) [reply rust/lance/src/index/vector/pq.rs:366]: need a flag for dot distance to not residualize

Comment (eddyxu) [reply rust/lance/src/index/vector/pq.rs:366]: Ok, checked it on the caller

**Pull request merged**